packages/assets-controllers/src/NftDetectionController.ts

```diff
@@ -300,6 +323,116 @@ export class NftDetectionController extends BaseController<
    */
   isMainnet = (): boolean => this.config.chainId === ChainId.mainnet;
 
+  isMainnetByNetworkClientId = (networkClient: NetworkClient): boolean => {
+    return networkClient.configuration.chainId === ChainId.mainnet;
+  };
+
+  #selectedAddressByNetworkClientId = new Map<NetworkClientId, string>();
+
+  override startPollingByNetworkClientId(
+    networkClientId: string,
+    selectedAddress?: string,
+  ): string {
+    // selectedAddress needs to be optional in the interface for the override to work with the mixin
+    if (!selectedAddress) {
+      throw new Error('no address specified');
+    }
+    this.#selectedAddressByNetworkClientId.set(
+      networkClientId,
+      selectedAddress,
+    );
+    return super.startPollingByNetworkClientId(networkClientId);
+  }
+
+  stopAllPolling(): void {
+    this.#selectedAddressByNetworkClientId.clear();
+    super.stopAllPolling();
+  }
+
+  async detectNftsByNetworkClientIdAndUserAddress(
```

Comment:
very similar in shape to `detectNfts`. Can they be DRY'ed a little? Maybe `detectNfts` have a overloaded definition instead. idk

Reply:
i think we want to eventually remove the original `detectNfts` when we fully switch over to using networkClientId for everything

Reply:
any reason we shouldn't make `networkClientId` and `selectedAddress` optional args and do light branching logic based on whether they were passed in? This is the pattern I've been using for other controller refactors (for example [here](https://github.com/MetaMask/core/pull/1763/files#diff-55d698549301c5a268bec6ce22d195a117bee3a060007c411cfa6f8aa6b032e8R249) and [here](https://github.com/MetaMask/core/pull/1698/files#diff-7086c0b9bbbb670051a9d0a84a0ab229731eecbe408914be00e1f0e00bebd71dR1291)) - I guess we did it the way you do it this PR for the gasFeeController. Either way works I suppose, and will ultimately be fixed and cleaned-up when we strip out the old way eventually but might be good for us to align on a pattern where possible.

Reply:
Devoid of any constraints, my vote would be to avoid introducing extra logic where we can. If network client ID is something that we are introducing across the board, it would be nice to not make `networkClientId` and `selectedAddress` optional, so that we don't have more scenarios we have to make sure we account for (we'd also have to make sure we test for it). On the other hand, if we can't guarantee that we will have a network client ID available when this is called, then it would make sense to make them optional. So I guess my question is, what was the original thought for making `networkClientId` optional in the PRs that were linked here?

Reply:
the idea here was that we want to didn't want to force the clients to have to immediately adopt this new pattern while the PollingController abstraction is still being refined and while a ton of breaking changes are already piled quite high for mobile. We didn't want any other critical core side changes to be stacked behind these breaking changes. Rather we've been aiming to make all of the multichain controller refactors "additive" and backwards compatible and will switch over to leverage the new pattern after the refactors are complete. Though as I say this I'm thinking we could/should consider doing these integrations on the extension sooner to make sure everything is working together.

Reply:
For sure. But good thinking to be cautious about it, too, so that you or the client teams are free to upgrade to these changes one at a time more slowly. Making these arguments optional seems like a good plan.

Reply:
Ok cool thanks for the feedback here Elliot. The question of making the `networkClientId` arg optional is a bit orthogonal to the question of a preferred pattern - new methods or added optional args - for making these refactors backwards compatible. Do you @mcmire or @jiexi or @BelfordZ have any particular thoughts on that point?

Reply:
Oh sorry, I misunderstood the question.

Working toward a set of cohesive shared libraries, I see two constraints here that come into play (there could be more, but these are the first two that come to mind). One is the desire for the APIs that we create to be as minimal as possible. This makes them easier to use and helps maintainability as well. On the other hand, we want to reduce the risk that in modifying our APIs we will introduce a breaking change.

Adding a new method isn't terrible on its own if we are essentially replacing one with another, because at least it satisfies the first constraint. However, this new method seems to be named after the arguments that it takes, and I feel like that's not as future-looking as it could be.

Given this, my two cents is that it would be better to keep our existing method and add options to it rather than make a new method. There would be a downside in that it would be more complicated to maintain, as we'd now have to make sure to test and keep track of multiple code paths, but I think that would be an acceptable compromise.

Reply:
> my two cents is that it would be better to keep our existing method and add options to it rather than make a new method.
This is my preferred pattern as well☝️